File: frontend/src/api/repairReport.ts
/**
 * 報修未完成報表 API Client
 * Prefix: /api/v1/repair-report
 */
import apiClient from '@/api/client'

// apiClient 的 baseURL 已設為 /api/v1，此處只需路徑後綴
const BASE = '/repair-report'

// ── 型別定義 ──────────────────────────────────────────────────────────────────

export interface UnifiedCase {
  source:           'hotel' | 'mall'
  source_label:     string
  ragic_id:         string
  case_no:          string
  occurred_at:      string | null
  floor:            string
  repair_type:      string
  title:            string
  status:           string
  responsible_unit: string
  pending_days:     number | null
  is_overdue:       boolean
  synced_at:        string | null
  finance_note:     string
  ragic_url:        string
}

export interface KpiData {
  total_unfinished:   number
  hotel_unfinished:   number
  mall_unfinished:    number
  overdue_count:      number
  avg_pending_days:   number
  max_pending_days:   number
  new_this_month:     number
  new_today:          number
}

export interface FilterOptions {
  statuses:     string[]
  repair_types: string[]
}

export interface UnfinishedCasesResponse {
  items:          UnifiedCase[]
  total:          number
  kpi:            KpiData
  filter_options: FilterOptions
}

export interface UnfinishedCasesParams {
  year:               number
  month:              number
  source?:            'all' | 'hotel' | 'mall'
  status_filter?:     string
  overdue_only?:      boolean
  repair_type_filter?: string
  keyword?:           string
  page?:              number
  page_size?:         number
}

export interface Recipient {
  id:         number
  name:       string
  email:      string
  department: string
  role:       string
  is_active:  boolean
  created_at: string | null
  updated_at: string | null
  created_by: string
  updated_by: string
}

export interface RecipientCreate {
  name:       string
  email:      string
  department?: string
  role?:      string
  is_active?: boolean
}

export interface RecipientUpdate {
  name?:       string
  email?:      string
  department?: string
  role?:       string
  is_active?:  boolean
}

export interface ScheduleSettings {
  id:                       number
  schedule_name:            string
  is_enabled:               boolean
  send_time:                string
  report_year_month_mode:   'current_month' | 'previous_month'
  include_hotel:            boolean
  include_mall:             boolean
  include_excel_attachment: boolean
  email_subject_template:   string
  email_body_template:      string
  created_at:               string | null
  updated_at:               string | null
  updated_by:               string
}

export interface ManualSendRequest {
  year:                    number
  month:                   number
  include_hotel?:          boolean
  include_mall?:           boolean
  include_excel_attachment?: boolean
  recipient_ids?:          number[]
}

export interface SendResult {
  recipient_email: string
  recipient_name:  string
  success:         boolean
  error_message:   string | null
}

export interface ManualSendResponse {
  sent_count:   number
  failed_count: number
  results:      SendResult[]
}

export interface MailLog {
  id:                     number
  send_date:              string
  send_time:              string
  report_year:            number
  report_month:           number
  recipient_email:        string
  recipient_name:         string
  subject:                string
  status:                 'success' | 'failed' | 'skipped'
  error_message:          string | null
  hotel_unfinished_count: number | null
  mall_unfinished_count:  number | null
  total_unfinished_count: number | null
  attachment_filename:    string | null
  created_at:             string | null
}

export interface MailLogListResponse {
  items: MailLog[]
  total: number
}

// ── API 函數 ──────────────────────────────────────────────────────────────────

export const repairReportApi = {
  /** 取得未完成案件（含 KPI + filter_options） */
  getUnfinishedCases: (params: UnfinishedCasesParams) =>
    apiClient.get<UnfinishedCasesResponse>(`${BASE}/unfinished-cases`, { params })
      .then(r => r.data),

  /** 匯出 Excel（產生下載 URL） */
  getExportUrl: (params: Omit<UnfinishedCasesParams, 'page' | 'page_size'>) => {
    const qs = new URLSearchParams(
      Object.fromEntries(
        Object.entries(params)
          .filter(([, v]) => v !== undefined && v !== null && v !== '')
          .map(([k, v]) => [k, String(v)])
      )
    ).toString()
    return `/api/v1${BASE}/export?${qs}`
  },

  // ── 收件人 ────────────────────────────────────────────────────────────────
  listRecipients: () =>
    apiClient.get<Recipient[]>(`${BASE}/recipients`).then(r => r.data),

  createRecipient: (payload: RecipientCreate) =>
    apiClient.post<Recipient>(`${BASE}/recipients`, payload).then(r => r.data),

  updateRecipient: (id: number, payload: RecipientUpdate) =>
    apiClient.put<Recipient>(`${BASE}/recipients/${id}`, payload).then(r => r.data),

  deleteRecipient: (id: number) =>
    apiClient.delete(`${BASE}/recipients/${id}`).then(r => r.data),

  testSendToRecipient: (id: number) =>
    apiClient.post<{ success: boolean; message: string }>(`${BASE}/recipients/${id}/test-send`)
      .then(r => r.data),

  // ── 排程設定 ──────────────────────────────────────────────────────────────
  getSchedule: () =>
    apiClient.get<ScheduleSettings>(`${BASE}/schedule`).then(r => r.data),

  updateSchedule: (payload: Partial<ScheduleSettings>) =>
    apiClient.put<ScheduleSettings>(`${BASE}/schedule`, payload).then(r => r.data),

  // ── 手動寄送 ──────────────────────────────────────────────────────────────
  sendNow: (payload: ManualSendRequest) =>
    apiClient.post<ManualSendResponse>(`${BASE}/send-now`, payload).then(r => r.data),

  // ── 寄送紀錄 ──────────────────────────────────────────────────────────────
  getMailLogs: (params?: {
    year?: number; month?: number; status?: string; email?: string;
    page?: number; page_size?: number
  }) =>
    apiClient.get<MailLogListResponse>(`${BASE}/mail-logs`, { params }).then(r => r.data),
}

File: backend/app/schemas/repair_report.py
"""
報修未完成報表 — Pydantic Schemas
"""
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Optional

from pydantic import BaseModel, EmailStr, field_validator


# ── 統一案件格式（前端顯示用）────────────────────────────────────────────────

class UnifiedCase(BaseModel):
    source:           str           # hotel / mall
    source_label:     str           # 飯店 / 商場
    ragic_id:         str
    case_no:          str
    occurred_at:      Optional[str]
    floor:            str
    repair_type:      str
    title:            str
    status:           str
    responsible_unit: str           # 工務處理人員（responsible_unit 欄位）
    pending_days:     Optional[int]
    is_overdue:       bool
    synced_at:        Optional[str]
    finance_note:     str
    ragic_url:        str


class KpiData(BaseModel):
    total_unfinished:   int
    hotel_unfinished:   int
    mall_unfinished:    int
    overdue_count:      int
    avg_pending_days:   float
    max_pending_days:   int
    new_this_month:     int
    new_today:          int


class FilterOptions(BaseModel):
    statuses:     list[str]
    repair_types: list[str]


class UnfinishedCasesResponse(BaseModel):
    items:          list[UnifiedCase]
    total:          int
    kpi:            KpiData
    filter_options: FilterOptions


# ── 收件人 ────────────────────────────────────────────────────────────────────

class RecipientBase(BaseModel):
    name:       str
    email:      EmailStr
    department: str = ""
    role:       str = ""
    is_active:  bool = True


class RecipientCreate(RecipientBase):
    pass


class RecipientUpdate(BaseModel):
    name:       Optional[str]  = None
    email:      Optional[EmailStr] = None
    department: Optional[str]  = None
    role:       Optional[str]  = None
    is_active:  Optional[bool] = None


class RecipientOut(RecipientBase):
    id:         int
    created_at: Optional[datetime]
    updated_at: Optional[datetime]
    created_by: str
    updated_by: str

    class Config:
        from_attributes = True


# ── 排程設定 ──────────────────────────────────────────────────────────────────

class ScheduleSettingsBase(BaseModel):
    schedule_name:            str  = "每日報修未完成報表"
    is_enabled:               bool = False
    send_time:                str  = "08:30"   # HH:MM
    report_year_month_mode:   str  = "current_month"
    include_hotel:            bool = True
    include_mall:             bool = True
    include_excel_attachment: bool = True
    email_subject_template:   str  = "【報修未完成報表】{year}年{month}月｜未完成 {total} 件｜飯店 {hotel_count} 件｜商場 {mall_count} 件"
    email_body_template:      str  = ""

    @field_validator("send_time")
    @classmethod
    def validate_send_time(cls, v: str) -> str:
        try:
            hour, minute = v.split(":")
            assert 0 <= int(hour) <= 23
            assert 0 <= int(minute) <= 59
        except Exception:
            raise ValueError("send_time 格式必須為 HH:MM（例如 08:30）")
        return v


class ScheduleSettingsUpdate(ScheduleSettingsBase):
    pass


class ScheduleSettingsOut(ScheduleSettingsBase):
    id:         int
    created_at: Optional[datetime]
    updated_at: Optional[datetime]
    updated_by: str

    class Config:
        from_attributes = True


# ── 手動寄送請求 ──────────────────────────────────────────────────────────────

class ManualSendRequest(BaseModel):
    year:                    int
    month:                   int
    include_hotel:           bool = True
    include_mall:            bool = True
    include_excel_attachment: bool = True
    recipient_ids:           list[int] = []   # 空 = 全部 active 收件人


class SendResult(BaseModel):
    recipient_email: str
    recipient_name:  str
    success:         bool
    error_message:   Optional[str]


class ManualSendResponse(BaseModel):
    sent_count:   int
    failed_count: int
    results:      list[SendResult]


# ── 寄送紀錄 ──────────────────────────────────────────────────────────────────

class MailLogOut(BaseModel):
    id:                    int
    send_date:             date
    send_time:             str
    report_year:           int
    report_month:          int
    recipient_email:       str
    recipient_name:        str
    subject:               str
    status:                str
    error_message:         Optional[str]
    hotel_unfinished_count:  Optional[int]
    mall_unfinished_count:   Optional[int]
    total_unfinished_count:  Optional[int]
    attachment_filename:   Optional[str]
    created_at:            Optional[datetime]

    class Config:
        from_attributes = True


class MailLogListResponse(BaseModel):
    items: list[MailLogOut]
    total: int


# ── 測試寄送 ──────────────────────────────────────────────────────────────────

class TestSendResponse(BaseModel):
    success:       bool
    message:       str
